lib/net.js

```diff
@@ -300,39 +299,42 @@ Socket.prototype._unrefTimer = function _unrefTimer() {
 
 function shutdownSocket(self, callback) {
   var req = new ShutdownWrap();
-  req.oncomplete = callback;
+  req.oncomplete = afterShutdown;
   req.handle = self._handle;
+  req.callback = callback;
   return self._handle.shutdown(req);
 }
 
 // the user has called .end(), and all the bytes have been
 // sent out to the other side.
-function onSocketFinish() {
+Socket.prototype._final = function(cb) {
   // If still connecting - defer handling 'finish' until 'connect' will happen
```

Comment:
Nit: I'd remove `handling 'finish'` as we are no longer handling the `'finish'` event.

Reply:
@lpinca done! And thanks for the reviews, it’s really helpful to talk this through with you!